### ISSUE TITLE
Please use Dart streams for Animation and other listeners

### ISSUE THREAD

**Issue by escamoteur** [flutter/flutter#15938]
Hi,
while working on an Animation want to remove my Listeners at the end of the Animation. Unfortunately you have to pass the Listener function to remove it, which makes it impossible to use with Lambda function.

See like in this example:

```dart
    _pageMoveToSubscription = theViewModel.scrollToEvents
         .listen((index) {
           animation.addStatusListener((status) {
             if (status == AnimationStatus.completed) {
               _pageController.animateToPage(index,
                                            curve: Curves.linear,
                                            duration: new Duration(milliseconds: 1));
               controller.reverse();
               animation.removeStatusListener(????) <<<<< see here
             }                                                    
             controller.forward();
           });
         });
```

It would be way easier if Animations would offer a Stream which you can listen on and which subscriptions you can easily cancel. Additionally you would get all the power of Stream manipulations.

**Comment by Hixie** [comment]
How about:

```dart
    _pageMoveToSubscription = theViewModel.scrollToEvents
         .listen((index) {
           void handler(status) {
             if (status == AnimationStatus.completed) {
               _pageController.animateToPage(index,
                                            curve: Curves.linear,
                                            duration: new Duration(milliseconds: 1));
               controller.reverse();
               animation.removeStatusListener(handler);
             }                                                    
             controller.forward();
           }
           animation.addStatusListener(handler);
         });
```

I haven't tested it but I'm pretty sure that works.

**Comment by escamoteur** [comment]
Still, where do you remove the StatusListener. Besides that if you would use Streams listeners we could use all Streams or Rx functions on the Data. Thinking on the proposed BLOCK pattern Streams would also make more sense than methods.
Dart has this beautifully feature, so why not using it?

**Comment by Hixie** [comment]
> Still, where do you remove the StatusListener.

I'm not sure what you mean. In the snippet I pasted, there's a call to `animation.removeStatusListener(handler);`, which removes the listener.

At a broader level, though, we mostly don't use Streams in Flutter's framework itself for a few reasons: Streams are quite heavy-weight (the API is very broad, compared to `Listenable`), the API is hard to extend (it's non-trivial to make your own streams and stream convertors and so on), and Streams are inherently asynchronous, which makes their use for sending data along a little inconvenient in the framework (unlike, say, `ValueListenable` which has a synchronous `value` property).

However, you can use `StreamBuilder` to use Streams with Flutter widgets, and there's no reason I can think of why you couldn't adapt from Stream to `Listenable` or `ValueListenable` and then use them throughout the framework.

**Comment by tudor07** [comment]
Please reconsider this, it is really painful to work with the current animations listeners api.

**Comment by Hixie** [comment]
Maybe we could create a `ValueListenable<T>`-to-`Stream<T>` adapter... If this is something that interests you, I recommend either writing a package that implements that, or filing a bug suggesting it.

**Comment by escamoteur** [comment]
Still it is pretty sad that a reactive framework doesn't go all the way down and uses Streams inside,

**Comment by tudor07** [comment]
Honestly for me it would be great if we would have at least a `clear()` method on the Animation so we can remove all listeners at once.

**Comment by Hixie** [comment]
@tudor07 That seems like a layering violation -- you can't necessarily know what objects are listening to the object, so removing the listeners may mean removing listeners that another object owns.

Can you elaborate on what you are doing that you end up wanting this feature?

**Comment by tudor07** [comment]
@Hixie 
What I want to do, is everytime I run an animation, it may start/end with different values, so I am creating a new Tween eveytime, and these Tweens have the same AnimationController as parent, so I guess the old listeners remain present. So my solution for this is:

1. Declare a `Function` on the `StatefulWidget` whici is the listener
2. Eveytime I want to run the animation I have to do:
```
// field in StatefulWidget class
Function _listener;

...

// in some method where I run the animation
    final tween = Tween<double>(begin: _someValu, end: _someOtherValue)
        .animate(CurvedAnimation(
      parent: _parentAnimController,
      curve: Curves.linear,
    ));

    tween.removeListener(_listener);
    _listener = () {
      // logic from my listener that changes the values so it animates
     setState(() => _val = tween.val);
    };
    tween.addListener(_listener);
```

It would be cool if there was a solution to not have the top-level `Function` listener, so I can run my listener only once and the second time I run the animation I don't have to remove and re-add.

But again, maybe I am doing something wrong in the first place, I would be happy to hear about a better solution.

Maybe for my case it would be nice to have a `addListenerOnce` method that removes the listener after it finished.

**Comment by escamoteur** [comment]
@Hixie I think that is a good solution!!! Thanks a lot

**Comment by shinayser** [comment]
> @tudor07 Not sure what you mean by "the old listeners remain present". That shouldn't be the case.
> But in any case that seems like a question for a different issue. This issue is about streams.
> 
> @escamoteur any opinions regarding the `valueListenableToStreamAdapter()` method from [master...Hixie:listenable_stream](https://github.com/flutter/flutter/compare/master...Hixie:listenable_stream)?

Hey, may I ask why those changes were not merged on `change_notifier.dart` yet? I might be crazy but I just can't find it.

My flutter version is the stable 1.7.8

**Comment by Hixie** [comment]
It wasn't proposed as a feature to put in Flutter, more as a proof-of-concept. You should totally feel free to take that source and make a package from it though! We could even put it in Flutter if it proves sufficiently popular.

**Comment by shinayser** [comment]
Amazing! The only thing I think could be a problem is that the compiler complains about "unclosed sinks" for the StreamController on the code.

Is this a real problem or can we ignore it ?
